Block metadata refactoring (part 2) - merging of fringes

tgrospic opened the pull request:
## Overview

This PR introduces calculation of parent pre-state with the use of the new multi-parent finalizer.

Major changes include:
- reorganization (simplification) of BlockMesasge fields, remove unnecessary nested objects and put all important fields in the root like block number, pre and post state hashes, deploys, etc.
- introduction of parents merged state object (replacement for Casper snapshot)
- introduction of new insert function in BlockDagStorage, with removed domain logic
- validation of blocks to include BlockMetadata in the result
- timestamp from BlockMessage is removed but it should be added later with modified behavior.

There are also many other changes related to cleanup the code which is unused or not usable with the new multi-parent finalizer.
There is also a lot of TODO comments which must be addressed in subsequent PR to reduce the scope because this PR has already become huge.

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

tgrospic committed:
Remove unused AtomicSyncVar and AtomicSyncVarF

tgrospic committed:
Migrate ConnectionCell to use Ref

tgrospic committed:
Remove Cell abstraction

tgrospic committed:
Replace MonadSTate with Ref in block metadata store

tgrospic committed:
Remove unused justification helper functions

tgrospic committed:
Use only block hash as justifications in BlockMetadata

tgrospic committed:
Use only block hash (remove validator hash) in justifications

tgrospic committed:
Fix genesis block sender to always be defined

tgrospic committed:
Fix latest messages, respect created blocks, seq num starts from 0

tgrospic committed:
Fix signing block without changing block sender

tgrospic committed:
Fix tests related to block sender

tgrospic committed:
Remove obsolete test for block dag storage (with empty sender)

tgrospic committed:
Remove obsolete (empty) test for vaults import from Ethereum

tgrospic committed:
Fix runtime manager compute genesis sender for mergeable store

tgrospic committed:
Move version field on BlockMessage to top level, remove in Header

tgrospic committed:
Change seqNum from Int to Long

tgrospic committed:
Remove unused extraBytes from block related protobuf types

tgrospic committed:
Removed obsolete and unused tests (RholangBuildTest)

tgrospic committed:
Remove block timestamp

tgrospic committed:
Remove parents from BlockMessage

tgrospic committed:
Remove JustificationProto type

tgrospic committed:
Move blockNumber to top level in BlockMessage

tgrospic committed:
Move bonds to top level in BlockMessage and change type to Map[Validator, Long]

tgrospic committed:
Move preStateHash to top level in BlockMessage

tgrospic committed:
Move postStateHash to top level in BlockMessage, remove RChainState

tgrospic committed:
Rearrange order of fields in BlockMessage, formatting

tgrospic committed:
Rename BodyProto to RholangStateProto in BlockMessage (state field)

tgrospic committed:
Move rejectedDeploys to top level in BlockMessage

tgrospic committed:
Add additional rejection fields to BlockMessage (blocks and senders)

tgrospic committed:
Fix block hashing function

tgrospic committed:
Initial test for block hashing

tgrospic committed:
Temporary change to casper snapshot to minimize failed tests

tgrospic committed:
Remove unused invalid format and version block errors

tgrospic committed:
Update LFS to receive finalized fringe data

tgrospic committed:
Fix LastFinalizedHeightConstraintChecker

tgrospic committed:
Parents fix in BlockMeatadataStore

tgrospic committed:
Fix block summary validation test

tgrospic committed:
Fix bonding validator test

tgrospic committed:
Fix SynchronyConstraintChecker for new finalizer

tgrospic committed:
Change CasperSnapshot to use fringe instead of LFB

tgrospic committed:
Adjust DAG merger to use fringe instead of LFB

tgrospic committed:
Fix parents filtered invalid blocks from justifications

tgrospic committed:
Update PyRChain to 0.4.0-alpha1 (new BLockMessage format)

tgrospic committed:
Fix genesis ceremony integration test

tgrospic committed:
Fix slashing integration test

tgrospic committed:
Fix VaultBalanceGetter test for correct balance in genesis block

tgrospic committed:
Remove duplicate code in genesis builder function for testing

tgrospic committed:
Move constants on top of GenesisBuilder object for testing

tgrospic committed:
Remove duplicated parameter RuntimeManager from BlockCreator.create

tgrospic committed:
Remove unused SynchronyConstraintChecker from NodeLaunch and NodeSyncing

tgrospic committed:
Remove unused LastFinalizedHeightConstraintChecker from NodeLaunch and NodeSyncing

tgrospic committed:
Remove ProtoUtil.deploys and replace with direct access to block deploys

tgrospic committed:
Remove ProtoUtil.systemDeploys and replace with direct access to block systemDeploys

tgrospic committed:
Move helper function bondToBondInfo from ProtoUtil to BlockApi

tgrospic committed:
Introduce BlockVersion object, fix version validation, add validation of failed deploys in genesis block

tgrospic committed:
Fix genesis block signature

tgrospic committed:
Approved block (now finalized fringe) cleanup (except ApprovedStore)

tgrospic committed:
Remove not needed Signature wrapper case class

tgrospic committed:
Remove unnecessary block validation BlockDagKeyValueStorage insert method

tgrospic committed:
Remove invalid blocks storage

tgrospic committed:
Update DagMessageState, removed creator reference, default constructor

tgrospic committed:
Fix LFS with separate blocks queue then receiver

tgrospic committed:
Remove invalid sender status and obsolete test for block sender

- block with invalid signatures are ignore before ever reaching validation

tgrospic committed:
Simplify BlockRetriever logic with received blocks

- fix fork-choice tips request on idle

tgrospic committed:
Use parametric effect in InterpreterUtil.mkTerm instead of Coeval

tgrospic committed:
Add BlockDagStorage to CI workflow

tgrospic committed:
Remove directlyFinalized flag in BlockMetadata

nzpr commented on models/src/main/protobuf/CasperMessage.proto:
```diff
@@ -79,12 +79,17 @@ message BlockMessageProto {
+  // Rejected data in finalized fringe
+  repeated bytes rejectedDeploys    = 11 [(scalapb.field).collection_type="List"];
+  repeated bytes rejectedBlocks     = 12 [(scalapb.field).collection_type="List"];
+  repeated bytes rejectedSenders    = 13 [(scalapb.field).collection_type="List"];
```
why do you want to put these 3 in a block? and what are `rejectedBlocks` and `rejectedSenders `?

tgrospic replied:
These fields store data for consensus algorithm and are related to the finalized fringe seen by the block. It's not necessary for the node to store this data in the block because it will have it in internal state when running consensus algorithm, but for LFS node which is starting from a fringe it will be used to create the next fringe after that.

1. **Rejected deploys** - deploys rejected in the fringe
2. **Rejected blocks** - blocks that are excluded from giving stake to finalized the fringe (invalid blocks)
3. **Rejected senders** - validators excluded from giving stake to finalized the fringe, but without having invalid e.g. lazy validator not producing new blocks

tgrospic committed:
Refactor RuntimeManager as implicit parameter

- unify validate block related functions to use RuntimeManager as implicit parameter
- remove unused Time from Validate.blockSummary and MultiParentCasper.validate
- remove unused Time from BlockGenerator test

tgrospic committed:
Remove completely CasperSnapshot from block validation

tgrospic committed:
Remove unused genesis block parameter from BlockGenerator.step test helper function

tgrospic committed:
Fix (quick) for arithmetic errors in reducer (transform to ReduceError)

tgrospic committed:
Remove Exception error from BlockStatus (errors), propagated as fatal error

tgrospic committed:
Remove SynchronyConstraintChecker and LastFinalizedHeightConstraintChecker

- remove integration test for synchrony constraint

tgrospic committed:
Remove DagRepresentation from CasperSnapshot

tgrospic committed:
Remove synchrony and LFB constraint tests

tgrospic committed:
Remove unused setup file in tests (already part of PR#3720)

tgrospic committed:
Update BlockMetadata and CasperSnapshot fields and remove unused

- remove fault-tolerance-threshold from config
- remove finalization-rate from config

tgrospic committed:
Remove unused empty constructor for DagRepresentation

tgrospic committed:
Remove block DAG storage tests

nzpr commented on block-storage/src/main/scala/coop/rchain/blockstorage/ApprovedStore.scala:
```diff
@@ -2,26 +2,27 @@ package coop.rchain.blockstorage
 
 object approvedStore {
-  type ApprovedStore[F[_]] = KeyValueTypedStore[F, Byte, ApprovedBlock]
+  type ApprovedStore[F[_]] = KeyValueTypedStore[F, Byte, FinalizedFringe]
```
It looks like this is not used at all? No calls on get on this KV store.
If this is supposed to be used in the process of LFS somehow - I think this can be just
`bootstrapFringe: KeyValueTypedStore[F, Unit, Set[BlockHash]],`  inside `BlockDagKeyValueStorage` or `BlockMetadataStore`

tgrospic replied:
ApprovedStore is used to store a fringe when downloaded as a starting point for LFS. I didn't want to change the name to reduce changes in many files.

Persisting of fringe is meant to be used to continue downloaded if node is interrupted in LFS process which is not implemented now so _get_ is not used.

nzpr commented on casper/src/main/resources/RevVault.rho:
```diff
@@ -206,8 +206,8 @@ in {
                       if (Nil != *logCh) {
                         new bd(`rho:block:data`), bdCh in {
                           bd!(*bdCh) |
-                          for (@blockNumber, @timestamp, @sender <- bdCh) {
```
Are we removing timestamp from the block permanently?

tgrospic replied:
No, we will introduce it again but with different validation and naming.

nzpr commented on casper/src/main/scala/coop/rchain/casper/blocks/proposer/BlockCreator.scala:
```diff
@@ -36,37 +37,41 @@ object BlockCreator {
-  def create[F[_]: Concurrent: Log: Time: BlockStore: BlockDagStorage: Metrics: RuntimeManager: Span](
-      s: CasperSnapshot,
+  def create[F[_]: Concurrent: Time: RuntimeManager: BlockDagStorage: BlockStore: Log: Metrics: Span](
+      preState: ParentsPreState,
```
I think it should be `ParentsPostState`

tgrospic replied:
Makes sense. To be even more precise I'll rename it to `ParentsMergedState`.

nzpr commented on casper/src/main/scala/coop/rchain/casper/dag/BlockDagKeyValueStorage.scala:
```diff
@@ -1,196 +1,96 @@
+                    val fringeStateHash       = blockMetadata.fringeStateHash.toBlake2b256Hash
+                    val fringeRejectedDeploys = block.rejectedDeploys.toSet
+                    val fringeStateRecord     = (fringeStateHash, fringeRejectedDeploys)
+                    val newFringes            = dr.fringeStates + ((msg.fringe, fringeStateRecord))
```
I'm thinking maybe instead of 
`fringeStates: Map[Set[BlockHash], (Blake2b256Hash, Set[ByteString])]`
to have 
`fringeStates: Map[Set[BlockHash], (Blake2b256Hash, Set[BlockHash])]` which contains blocks finalized compared to previous fringe, and separately 
`finalRejections: Map[BlockHash, Set[ByteString]]`

tgrospic replied:
I've added `fringeStates` temporary just to be able to get previous fringe state when calculating parents merged state. Maybe at the end we need to persist some of the data on disk or expand this cache to store the whole parents state result.

nzpr commented on casper/src/test/scala/coop/rchain/casper/batch2/RholangBuildTest.scala:
```diff
@@ -1,80 +0,0 @@
-import org.scalatest.flatspec.AnyFlatSpec
-import org.scalatest.matchers.should.Matchers
-
-class RholangBuildTest extends AnyFlatSpec with Matchers {
```
I'm curious what this test has been testing

tgrospic replied:
It's difficult to decipher all the nonsense I seen in _unit_ tests. This is just another example where description of the test doesn't correspond with what's testing.
So in this test, it's just repeats what is done in most of the tests so it doesn't contribute anything.

There are a few major categories of tests that doesn't make sense:
- it's not a unit test because it's testing overall behavior of the node (most of the tests)
- it's testing functionality not important for the tests or completely missing the testing subject
  - e.g. [GenesisTest](https://github.com/rchain/rchain/blob/15d631623e1f6d10ec53c306338541a5402c61a4/casper/src/test/scala/coop/rchain/casper/genesis/GenesisTest.scala) which is testing parsing of bonds file which is not related to genesis tests
  - BlockDagStorageTest (removed is this PR) which tests data stored on disk which is not part of storage functionality
- it's testing very specific data like data in the block but test build much more data not relevant for testing
  - e.g. [ValidateTest](https://github.com/rchain/rchain/blob/15d631623e1f6d10ec53c306338541a5402c61a4/casper/src/test/scala/coop/rchain/casper/batch2/ValidateTest.scala) is building the DAG to just check the functions operating on the block data

tgrospic committed:
Remove latest messages and LFB from BlockMetadataStore, use DagMessageState in DagRepresentation

tgrospic committed:
Add DagMerger.merge function for merging fringes

tgrospic committed:
Fix block receiver and validation, preparation for Casper snapshot refactoring

tgrospic committed:
Add ParentsPreState instead of Casper snapshot, new insert function in DAG storage

tgrospic committed:
Fix tests related to parent pre-state calculation

tgrospic committed:
Rename ParentsPreState to ParentsMergedState to be more clear what it represents

tgrospic committed:
Fix genesis integration test to check for read-only node

tgrospic commented on block-storage/src/main/scala/coop/rchain/blockstorage/dag/DagRepresentation.scala:
```diff
@@ -60,42 +67,20 @@ final case class DagRepresentation(
       validator: Validator
   ): F[Option[BlockHash]] =
-    dr.latestMessagesHashes.get(validator).pure
+    dr.dagMessageState.latestMsgs.filter(_.sender == validator).map(_.id).headOption.pure
```
With invalid blocks, validator can have multiple latest messages. We need to see how to manage this situation.

nzpr replied:
I agree - if you remember I was saying we need latest messages be Map[Sender, Set[Message]]. And we can actually merge multiple latest messages. We can punish for that, but we still can merge.

tgrospic replied:
Here I'm thinking about places where latest message is not used like BlockApi where external code can depend on this change.
https://github.com/rchain/rchain/blob/d9cd6dde980d22dff9a96c581dbe5d7d5d7b18a2/casper/src/main/scala/coop/rchain/casper/api/BlockApiImpl.scala#L655-L661

The Map you mentioned is practically there in DAG message state.
```scala
dag.dagMessageState.latestMsgs.groupBy(_.id) : Map[BlockHash, Set[Message[BlockHash, Validator]]]
```

tgrospic replied:
I've added a cleanup of latest message extensions from DagRepresentation state https://github.com/rchain/rchain/pull/3688/commits/6f7ba7ddf0c42fe2c9157954046ce0d2028c6970.

tgrospic commented on casper/src/main/scala/coop/rchain/casper/engine/NodeSyncing.scala:
```diff
@@ -115,57 +94,41 @@ class NodeSyncing[F[_]
-              s"Approved state for block ${PrettyPrinter.buildString(block, short = true)} is successfully restored."
-            )
+        _ <- ApprovedStore[F].putApprovedBlock(fringe)
+        _ <- LastApprovedBlock[F].set(fringe)
```
LastApprovedBlock can be removed now. EDIT: removed.

tgrospic committed:
Remove unused LastApprovedBlock in-memory state for LFB

tgrospic commented on casper/src/main/scala/coop/rchain/casper/protocol/CommUtil.scala:
```diff
@@ -138,14 +138,14 @@ final class CommUtilOps[F[_]](
       bootstrap      <- maybeBootstrap.liftTo(StandaloneNodeSendToBootstrapError)
-      msg            = ApprovedBlockRequest("", trimState).toProto
-      _              <- commUtil.sendWithRetry(ToPacket(msg), bootstrap, 10.seconds, "ApprovedBlockRequest")
+      msg            = FinalizedFringeRequest("", trimState).toProto
```
I don't know what's the purpose of _identifier_ on ApprovedBlockRequest (FinalizedFringeRequest). It's filled with emoty string and should probably be removed.

tgrospic commented on casper/src/main/scala/coop/rchain/casper/rholang/InterpreterUtil.scala:
```diff
@@ -47,70 +47,169 @@ object InterpreterUtil {
-      block: BlockMessage,
-      s: CasperSnapshot,
-      runtimeManager: RuntimeManager[F]
+  def validateBlockCheckpoint[F[_]: Concurrent: Timer: RuntimeManager: BlockDagStorage: BlockStore: Log: Metrics: Span](
```
validateBlockCheckpoint and validateBlockCheckpointNew are almost identical. The new version includes BlockMetadata in output. The old version is used is tests and should be removed when tests are fixed.

tgrospic commented on casper/src/test/scala/coop/rchain/casper/addblock/MultiParentCasperAddBlockSpec.scala:
```diff
@@ -201,27 +200,6 @@ class MultiParentCasperAddBlockSpec extends AnyFlatSpec with Matchers with Inspe
-      } yield (status shouldBe Left(InvalidSender))
-    }
-  }
-
```
New tests must be written for block validation logic.

tgrospic commented on casper/src/test/scala/coop/rchain/casper/batch2/ValidateTest.scala:
```diff
@@ -599,48 +498,45 @@ class ValidateTest
+      val actualValid = Validate.version[Task](blockWithVersion).runSyncUnsafe()
+
+      actualValid shouldBe expectedValid
+    }
```
Most of the validation tests should be rewritten as generative tests like these without creation of additional data not necessary for testing.

tgrospic committed:
Refactor old latest message method on DagRepresentation state, cleanup extensions

zsluedem commented on casper/src/main/scala/coop/rchain/casper/CasperSnapshot.scala:
```diff
@@ -1,51 +1,29 @@
-
-final case class OnChainCasperState(
-    shardConf: CasperShardConf,
+final case class ParentsMergedState(
```
Rename this file to `ParentsMergedState.scala`?

tgrospic replied:
I'm sustaining of file renaming/moving to reduce possible conflicts. I'll create a separate PR for reorganizing files in casper project.
Also I want to merge this PR first and leave your PR #3720 to resolve conflicts.

zsluedem commented on casper/src/main/scala/coop/rchain/casper/MultiParentCasper.scala:
```diff
@@ -59,182 +55,205 @@ object MultiParentCasper {
+
+      prevFringeStr      = PrettyPrinter.buildString(prevFringeHashes)
+      prevFringeStateStr = PrettyPrinter.buildString(prevFringeState.toByteString)
+      _                  = println(s"PREV FRINGE $prevFringeStr")
```
`Log` instead of `println`

tgrospic replied:
It's left over of debugging, I'll remove it.

zsluedem commented on casper/src/main/scala/coop/rchain/casper/MultiParentCasper.scala:
```diff
@@ -59,182 +55,205 @@ object MultiParentCasper {
+
+                            fringeBlocksCount = fringeBlocks.size
+                            fringeDiffBlocks  = PrettyPrinter.buildString(fringeBlocks)
+                            _                 = println(s"FRINGE DIFF ($fringeBlocksCount) $fringeDiffBlocks")
```
`Log` instead of `println`

tgrospic replied:
It's left over of debugging, I'll remove it.

zsluedem commented on rholang/src/main/scala/coop/rchain/rholang/interpreter/merging/RholangMergingLogic.scala:
```diff
@@ -202,7 +202,7 @@ object RholangMergingLogic {
-  val codecMergeableKey: Codec[(ByteVector, ByteVector, Int)] =
-    (codecByteVectorVar :: codecByteVectorVar :: uint16).as[(ByteVector, ByteVector, Int)]
+  val codecMergeableKey: Codec[(ByteVector, ByteVector, Long)] =
+    (codecByteVectorVar :: codecByteVectorVar :: vlong).as[(ByteVector, ByteVector, Long)]
```
A fix for https://discord.com/channels/391435113415573504/488440523326226433/989776507637477436 ?

tgrospic replied:
Yes, it's indirectly a fix, but caused by the change of seq number from Int to Long.

tgrospic committed:
Remove unused codecs

tgrospic committed:
Remove leftover println from debugging

nzpr reviewed (approved):
LGTM

stanislavlyalin commented on casper/src/main/scala/coop/rchain/casper/engine/NodeSyncing.scala:
```diff
@@ -227,59 +182,41 @@ class NodeSyncing[F[_]
       _ <- heightMap.flatMap(_._2).toList.reverse.traverse_ { hash =>
             for {
               block <- BlockStore[F].getUnsafe(hash)
+              // TODO: blocks added to DAG without validation will have flag `processed=false` so invalid flag is not applicable
```
Does this `TODO` imply some kind of action? It looks like it just explains why the flag is not needed

tgrospic replied:
It's both. We need to refactor LFS bootstrap procedure to be able to support invalid blocks so this code must be changed anyway.

stanislavlyalin commented on casper/src/test/scala/coop/rchain/casper/api/BlocksResponseAPITest.scala:
```diff
@@ -48,56 +48,40 @@ class BlocksResponseAPITest
              v2,
              bonds,
-             HashMap(v1 -> genesis.blockHash, v2 -> genesis.blockHash, v3 -> genesis.blockHash)
+             Seq(genesis.blockHash)
```
Is it using one hash instead of three because they are equal?

tgrospic replied:
Yes. Now justifications are completely correct, only validators which created a block are referenced instead of cheating and using the same block as created with different validators.

So first correction is to fill in the sender even for genesis block and second to have only one block as justification if building on top of genesis.

stanislavlyalin commented on models/src/main/scala/coop/rchain/models/BlockVersion.scala:
```diff
@@ -0,0 +1,16 @@
+    *  - previous versions are supported in transition period from previous to current version
+    *  - version 1 is still supported to validate previous blocks correctly
+    */
+  val Supported = List(Current)
```
What is the procedure for evaluating version compatibility?

tgrospic replied:
This should completely depend on specific change to the block structure or hashing of the block.

The reason to have list of supported versions is to be able to upgrade the node and validate previous block version until the whole network (required for validators) is on updated version.

This is minimum transition but depending on the case multiple versions can be supported to enable longer compatibility.

stanislavlyalin reviewed (approved):
Great work!

It is good to clean the PR, because when looking at commits step by step, some changes in one commit eliminate changes in another commit. But I understand it's a job that takes a lot of time.

It's amazing how much extra stuff was in the codebase!

zsluedem reviewed (approved):
LGTM

tgrospic commented:
bors merge :rocket:
